File: 04.py
#!/usr/bin/env python3.11


p1 = 0
p2 = 0

def parse(s):
    a, b = s.split('-')
    return int(a), int(b)

for line in open('04.txt'):
    x, y = map(parse, line.split(','))
    if (x[0]>=y[0] and x[1]<=y[1]) or (x[0]<=y[0] and x[1]>=y[1]):
        p1 += 1
    x, y = range(x[0], x[1]+1), range(y[0], y[1]+1)
    if set(x) & set(y):
        p2 += 1

print(p1, p2)

File: 04.hs
import Data.List
import Data.List.Split (splitOn)

parse :: String -> [[[Int]]]
parse = map (map pair . splitOn ",") . lines
    where
        interval [a, b] = [read a .. read b]
        pair = interval . splitOn "-"

subset :: [[Int]] -> Bool
subset l = foldl1 intersect l `elem` l

overlap :: [[Int]] -> Bool
overlap = not . null . foldl1 intersect

main = do
    inp <- parse <$> readFile "04.txt"
    print $ length . filter (== True) $ map subset inp
    print $ length . filter (== True) $ map overlap inp
